File: server/crates/application/src/queries/catalog.rs
use domain::{
    book::{port::BookReadRepoPort, Book},
    book_copy::{port::BookCopyReadRepoPort, BookCopy},
};
use std::sync::Arc;

use super::QueryError;

#[derive(Clone)]
pub struct CatalogQueries {
    book_read_repo: Arc<dyn BookReadRepoPort>,
    book_copy_read_repo: Arc<dyn BookCopyReadRepoPort>,
}

impl CatalogQueries {
    #[must_use]
    pub fn new(
        book_read_repo: Arc<dyn BookReadRepoPort>,
        book_copy_read_repo: Arc<dyn BookCopyReadRepoPort>,
    ) -> Self {
        Self {
            book_read_repo,
            book_copy_read_repo,
        }
    }

    pub async fn get_book_catalog(&self) -> Result<Vec<Book>, QueryError> {
        self.book_read_repo
            .get_catalog()
            .await
            .map_err(QueryError::from)
    }

    pub async fn get_book_by_isbn(&self, isbn: &str) -> Result<Option<Book>, QueryError> {
        self.book_read_repo
            .get_by_isbn(isbn)
            .await
            .map_err(QueryError::from)
    }

    pub async fn get_book_copy_details(
        &self,
        barcode: &str,
    ) -> Result<Option<BookCopy>, QueryError> {
        self.book_copy_read_repo
            .get_by_barcode(barcode)
            .await
            .map_err(QueryError::from)
    }
}

File: servers/go/internal/application/queries/catalog.go
package queries

import (
	"context"
	"fmt"

	"github.com/christophercaldwell/model-architecture/go/internal/domain/book"
	"github.com/christophercaldwell/model-architecture/go/internal/domain/bookcopy"
)

type CatalogQueries struct {
	bookRepo     book.ReadRepository
	bookCopyRepo bookcopy.ReadRepository
}

func NewCatalogQueries(bookRepo book.ReadRepository, bookCopyRepo bookcopy.ReadRepository) *CatalogQueries {
	return &CatalogQueries{
		bookRepo:     bookRepo,
		bookCopyRepo: bookCopyRepo,
	}
}

func (q *CatalogQueries) GetBookCatalog(ctx context.Context) ([]book.Book, error) {
	books, err := q.bookRepo.GetCatalog(ctx)
	if err != nil {
		return nil, fmt.Errorf("get book catalog: %w", err)
	}
	return books, nil
}

func (q *CatalogQueries) GetBookByISBN(ctx context.Context, isbn string) (*book.Book, error) {
	b, err := q.bookRepo.GetByISBN(ctx, isbn)
	if err != nil {
		return nil, fmt.Errorf("get book by isbn: %w", err)
	}
	return b, nil
}

func (q *CatalogQueries) GetBookCopyDetails(ctx context.Context, barcode string) (*bookcopy.BookCopy, error) {
	bc, err := q.bookCopyRepo.GetByBarcode(ctx, barcode)
	if err != nil {
		return nil, fmt.Errorf("get book copy details: %w", err)
	}
	return bc, nil
}
